kv.go

```diff
@@ -0,0 +1,436 @@
+	KeyValue(bucket string) (KeyValue, error)
+	// CreateKeyValue will create a KeyValue store with the following configuration.
+	CreateKeyValue(cfg *KeyValueConfig) (KeyValue, error)
+	// DeleteKeyValue will delete this KeyValue store (JetStream stream).
+	DeleteKeyValue(bucket string) error
+}
+
+type KeyValue interface {
+	// Get returns the latest value for the key.
+	Get(key string) (entry KeyValueEntry, err error)
+	// Put will place the new value for the key into the store.
+	Put(key string, value []byte) (revision uint64, err error)
+	// Create will add the key/value pair iff it does not exist.
+	Create(key string, value []byte) (revision uint64, err error)
+	// Update will update the value iff the latest revision matches.
+	Update(key string, value []byte, last uint64) (revision uint64, err error)
+	// Delete the key and all revisions.
+	Delete(key string) error
+	// WatchAll will invoke the callback for all updates.
+	WatchAll(cb KeyValueUpdate) (*Subscription, error)
+	// Watch will invoke the callback for any keys that match keyPattern when they update.
+	Watch(keys string, cb KeyValueUpdate) (*Subscription, error)
+	// List will return all values for the key.
+	List(key string) ([]KeyValueEntry, error)
+	// Bucket returns the current bucket name.
+	Bucket() string
+}
+
+// KeyValueConfig is for configuring a KeyValue store.
+type KeyValueConfig struct {
+	Bucket       string
+	Description  string
+	MaxValueSize int32
+	History      uint8
+	TTL          time.Duration
+	MaxBytes     int64
+	Storage      StorageType
+	Replicas     int
+}
+
+// Used to watch all keys.
+const (
+	AllKeys = ">"
+	kvop    = "KV-Operation"
+	kvdel   = "DEL"
+)
+
+type KeyValueOp uint8
+
+const (
+	KeyValuePut KeyValueOp = iota
+	KeyValueDelete
+)
+
+// Retrieved entry for Get or List or Watch.
+type KeyValueEntry interface {
+	// Bucket is the bucket the data was loaded from.
+	Bucket() string
+	// Key is the key that was retrieved.
+	Key() string
+	// Value is the retrieved value.
+	Value() []byte
+	// Revision is a unique sequence for this value.
+	Revision() uint64
+	// Created is the time the data was put in the bucket.
+	Created() time.Time
+	// Delta is distance from the latest value.
+	Delta() uint64
+	// Operation returns Update or Delete
+	Operation() KeyValueOp
+}
+
+// Callback handler for KeyValue updates.
+type KeyValueUpdate func(v KeyValueEntry)
+
+// Errors
+var (
+	ErrBucketNameRequired = errors.New("nats: bucket name is required")
+	ErrInvalidBucketName  = errors.New("nats: invalid bucket name")
+	ErrBucketNotFound     = errors.New("nats: bucket not found")
+	ErrBadBucket          = errors.New("nats: bucket not valid key-value store")
+)
+
+const (
+	kvBucketNameTmpl  = "KV_%s"
+	kvSubjectsTmpl    = "$KV.%s.>"
+	kvSubjectsPreTmpl = "$KV.%s."
+	kvNoPending       = "0"
+)
+
+// KeyValue will lookup and bind to an existing KeyValue store.
+func (js *js) KeyValue(bucket string) (KeyValue, error) {
```

Comment:
I think the point was to be consistent though no? I did add in a sanitizeName helper for ObjectStore which we could also use here.

Reply:
indeed, but you should be able to bind to any properly prepared stream (that was your request), and I think for replicas being specifically read only (being mirrors, typically) we would want to bind to them by specific name rather than have someone accidently access them as if they are a full on bucket.

Reply:
I agree originally I thought you should be able to free-form KV and ObjectStores and we would *allow* them, but do not think that now. Could be a support nightmare imo.

Read-Replicas and all of that need a bunch more thought around it so not touching that here. Also is applicable to more than just KV.

Reply:
ok, anyway we can add later if its a vararg it should at least not break too much :)

Reply:
Yes agree, wanted to try to avoid duplicating the JSOpt so went with binding to JS which solves that, now we can add in KVOpt or ObjOpts if needed but prefer to see how far we can get without them.